File: Intermediate/24.Solve the 01 Knapsack Problem/Solution.py
def knapsack(n, W, weights, values):
    dp = [[0] * (W + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for w in range(W + 1):
            if weights[i - 1] <= w:
                dp[i][w] = max(values[i - 1] + dp[i - 1][w - weights[i - 1]], dp[i - 1][w])
            else:
                dp[i][w] = dp[i - 1][w]

    return dp[n][W]


if __name__ == "__main__":
    n = int(input("Enter the number of items: "))
    weights = list(map(int, input("Enter the weights of the items: ").split()))
    values = list(map(int, input("Enter the values of the items: ").split()))
    W = int(input("Enter the maximum capacity of the knapsack: "))

    print("Maximum value:", knapsack(n, W, weights, values))

File: Intermediate/24.Solve the 01 Knapsack Problem/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

int knapsack(int n, int W, vector<int>& weights, vector<int>& values) {
    vector<vector<int>> dp(n + 1, vector<int>(W + 1, 0));

    for (int i = 1; i <= n; i++) {
        for (int w = 0; w <= W; w++) {
            if (weights[i - 1] <= w)
                dp[i][w] = max(values[i - 1] + dp[i - 1][w - weights[i - 1]], dp[i - 1][w]);
            else
                dp[i][w] = dp[i - 1][w];
        }
    }

    return dp[n][W];
}

int main() {
    int n, W;

    cout << "Enter the number of items: ";
    cin >> n;

    vector<int> weights(n), values(n);

    cout << "Enter the weights of the items:\n";
    for (int i = 0; i < n; i++)
        cin >> weights[i];

    cout << "Enter the values of the items:\n";
    for (int i = 0; i < n; i++)
        cin >> values[i];

    cout << "Enter the maximum capacity of the knapsack: ";
    cin >> W;

    cout << "Maximum value: " << knapsack(n, W, weights, values) << endl;

    return 0;
}
